File: 0735-asteroid-collision/0735-asteroid-collision.cpp

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> st;

        for (int i = 0; i < asteroids.size(); i++) {
            bool destroyed = false;

            while (!st.empty() && st.top() > 0 && asteroids[i] < 0) {
                // Handle collision
                if (abs(st.top()) < abs(asteroids[i])) {
                    st.pop(); // Top asteroid is destroyed
                } else if (abs(st.top()) == abs(asteroids[i])) {
                    st.pop(); // Both are destroyed
                    destroyed = true; // Current asteroid also destroyed
                    break;
                } else {
                    destroyed = true; // Current asteroid is destroyed
                    break;
                }
            }

            // Push the current asteroid if it survived
            if (!destroyed) {
                st.push(asteroids[i]);
            }
        }

        // Collect the result from the stack
        vector<int> ans;
        while (!st.empty()) {
            ans.push_back(st.top());
            st.pop();
        }
        reverse(ans.begin(), ans.end()); // Reverse to get the correct order
        return ans;
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] nums) {
        Stack<Integer> st = new Stack<>();
        st.add(nums[0]);
        for(int i=1;i<nums.length;i++){
           int num = nums[i];
           boolean des = false;
           while(st.size() > 0 && st.peek() > 0 && num < 0){
               if(st.peek() < -num) st.pop();
               else if(st.peek() == -num){
                st.pop();
                des = true;
                break;
               }
               else if(st.peek() > -num){
                des = true;
                break;
               }
           }
           if(des == false){
            st.add(num);
           }

          }
        

     int[] ans = new int[st.size()];
     int idx =st.size()-1;
     while(st.size() > 0){
      ans[idx--] = st.pop();
     }
     return ans;
    }
}